fix(GH-3684): aggregate CloudIntegrationEvents for connector task executions

igdianov opened the pull request:
This PR improves handling of CloudIntegrationEvents to ensure they are aggregated inside engine transaction context to be able to populate execution context headers in the audit message after transaction closes. The following events have been refactored:

- [x] CloudIntegrationRequestedEvent
- [x] CloudIntegrationResultReceivedEvent
- [x] CloudIntegrationErrorReceivedEvent

Part of https://github.com/Activiti/Activiti/issues/3684

igdianov committed:
fix(GH-3674): aggregate CloudIntegrationRequestedEvent during connector task execution

igdianov committed:
fix: update ConnectorAuditProducerIT test

igdianov committed:
fix: Add CompositeCommand base class

igdianov committed:
fix: add integration context event aggregation in transaction

igdianov committed:
fix: update unit tests

igdianov committed:
fix: resolve optimistic locking exception error

igdianov committed:
fix: correct integration error receive due to breaking AuditConnectorIT

igdianov committed:
fix: refactor ServiceTaskIntegrationResultEventHandlerTest mocks

igdianov committed:
fix: add required headers asserts in ConnectorAuditProducerIT

igdianov committed:
fix: removed deprecated sendAuditEvent method in IntegrationRequestSender

igdianov committed:
fix: correct Codaxy issues

igdianov committed:
fix: remove unused code

igdianov committed:
fix: preserve integration context event order

igdianov committed:
Revert "fix: preserve integration events aggregation order"

This reverts commit 1a8b104fa92257a1596ef42aff73bf776b939446.

igdianov committed:
fix: preserve integration events aggregation order

igdianov committed:
fix: defer add integration error event on closing execution context

igdianov committed:
fix: add CommandContextCloseListenerAdapter

igdianov committed:
fix: reduce CompositeCommand visibility to package

erdemedeiros commented on activiti-cloud-runtime-bundle-service/activiti-cloud-services-runtime-bundle/activiti-cloud-services-connectors/src/main/java/org/activiti/services/connectors/channel/CompositeCommand.java:
```diff
@@ -0,0 +1,32 @@
+
+import java.util.ArrayList;
+
+class CompositeCommand extends ArrayList<Command<?>> implements Command<Void> {
```
I'd prefer not to have `CompositeCommand` extending `ArrayList`, I'd rather receive a list of commands in the constructor and maybe add a utility method to initialize it:
```java
CompositeCommand.of(new Command1(), new Command2())
```

igdianov replied:
fixed

igdianov replied:
sounds good

erdemedeiros commented on activiti-cloud-runtime-bundle-service/activiti-cloud-services-runtime-bundle/activiti-cloud-services-connectors/src/main/java/org/activiti/services/connectors/channel/ServiceTaskIntegrationErrorEventHandler.java:
```diff
@@ -94,17 +83,10 @@ public void receive(IntegrationError integrationError) {
-                                                                                            execution));
+                            managementService.executeCommand(new TriggerIntegrationContextErrorCmd(integrationError,
+                                                                                                   execution));
+                            return;
```
I was wondering if we can avoid this `return`.

erdemedeiros reviewed (changes requested):
I have some concerns about `CompositeCommand`. LGTM otherwise.

erdemedeiros committed:
Extract commands and refactor test
- extract commands in dedicated classes to make them easier to test
- refactor ServiceTaskIntegrationErrorEventHandlerTest: as it's a unit test it should not test code that is out of the boundaries of ServiceTaskIntegrationErrorEventHandler

igdianov committed:
fix: correct license header

igdianov committed:
fix: correct Codaxy errors

igdianov committed:
fix: remove mockito-inline dependency

igdianov committed:
Revert "fix: remove mockito-inline dependency"

This reverts commit 1ff2bcd4a1947e532bc5840a24cc088f1002819b.
